### PR TITLE
add non-conservative convection term

### PR DESCRIPTION
**Pull request by qiauil**
* add non-conservative convection term
* allow to swift between conservative convection and non-conservative convection in steppers

### CODE REVIEW

File: exponax/nonlin_fun/_convection.py [hunk 1]
```diff
@@ -37,19 +39,41 @@ def __init__(
         channels as spatial dimensions and then gives
 
         ```
-            𝒩(u) = - b₁ 1/2 ∇ ⋅ (u ⊗ u)
+            𝒩(u) = -b₁ u ⋅ ∇ u
         ```
-
+        
         with `∇ ⋅` the divergence operator and the outer product `u ⊗ u`.
```
Comment (Ceyron) [inline exponax/nonlin_fun/_convection.py:45]: This part of the docstring is now in the wrong place. Could you fix this?

File: exponax/nonlin_fun/_convection.py [hunk 2]
```diff
@@ -37,19 +39,41 @@ def __init__(
...
+
+        Meanwhile, if you use a conservative form, the convection term is given by
+
+        ```
+            𝒩(u) = -b₁ u (u)ₓ
```
Comment (Ceyron) [inline exponax/nonlin_fun/_convection.py:50]: I thought we wanted to have the conservative form as 1/2 (u^2)_x in 1d?

File: exponax/nonlin_fun/_convection.py [hunk 3]
```diff
@@ -37,19 +39,41 @@ def __init__(
...
+        ```
+        
+        for 1D and
+
+        ```
+            𝒩(u) = -b₁ ∇ ⋅ (u ⊗ u)
```
Comment (Ceyron) [inline exponax/nonlin_fun/_convection.py:56]: By the conclusion we made in the issues thread, there should be a 1/2 here

File: exponax/nonlin_fun/_convection.py [hunk 4]
```diff
@@ -37,19 +39,41 @@ def __init__(
...
 
         ```
-            𝒩(u) = - b₁ 1/2 (1⃗ ⋅ ∇)(u²)
+            𝒩(u) = -b₁ 1/2 (1⃗ ⋅ ∇)(u²)
+        ```
+        
+        for the conservative form and
+        
+        ```
+            𝒩(u) = -b₁ 1/2 u (1⃗ ⋅ ∇)u
```
Comment (Ceyron) [inline exponax/nonlin_fun/_convection.py:71]: non-conservative single-channel should not have the 1/2, right? (cf table in https://github.com/Ceyron/exponax/issues/43#issuecomment-2421861564)

File: exponax/nonlin_fun/_convection.py [hunk 5]
```diff
@@ -101,24 +127,61 @@ def _multi_channel_eval(
...
+                "Number of channels in u_hat should match number of spatial dimensions"
+            )
+        u_hat_dealiased = self.dealias(u_hat)
+        u = self.ifft(u_hat_dealiased)
+        nabla_u = self.ifft(self.derivative_operator[None, :] * u_hat[:, None])
+        conv_u = jnp.sum(
+            u[None, :] * nabla_u,
+            axis=1,
+        )
+        #conv_u=sum(
```
Comment (Ceyron) [inline exponax/nonlin_fun/_convection.py:170]: Can you remove the commented piece of code

File: exponax/nonlin_fun/_convection.py [hunk 6]
```diff
@@ -142,10 +205,44 @@ def _single_channel_eval(
...
 
+    def _single_channel_nonconservative_eval(
+        self, u_hat: Complex[Array, "C ... (N//2)+1"]
+    ) -> Complex[Array, "C ... (N//2)+1"]:
+        """
+        Evaluates the non-conservative convection term for a single-channel state `u_hat` in
+        Fourier space. The convection term is given by
+
+        ```
+            𝒩(u) = -b₁ 1/2 u (1⃗ ⋅ ∇)u
```
Comment (Ceyron) [inline exponax/nonlin_fun/_convection.py:216]: non-conservative single-channel mode should be without the 1/2

Comment (Ceyron) [reply exponax/nonlin_fun/_convection.py:216]: (also reflects what happens in the code below)

File: exponax/stepper/_korteweg_de_vries.py [hunk 7]
```diff
@@ -19,6 +19,7 @@ class KortewegDeVries(BaseStepper):
     advect_over_diffuse: bool
     diffuse_over_diffuse: bool
     single_channel: bool
+    conservaitve: bool
```
Comment (Ceyron) [inline exponax/stepper/_korteweg_de_vries.py:22]: typo

**Review by Ceyron — CHANGES_REQUESTED** [review]
Hi Qiang,

thanks for the updated PR. I left some remarks in the code.

**Comment by Ceyron** [comment]
A couple more general remarks:

Can you run the pre-commit hook, being in the same folder as the repo and then:
```bash
pip install pre-commit
pre-commit install --install-hooks
pre-commit run --all-files
```

This should resolve code-linting checks.

And please keep docstrings to 80 characters in length if possible

**Comment by Ceyron** [comment]
Thanks for the updates, some last stylistic requests:

1. In the `_convection.py` file: please add a newline under the `**Arguments:**` header, otherwise mkdocs will not correctly produce the documentation
2. Please use [rewrap](https://marketplace.visualstudio.com/items?itemName=stkb.rewrap) to ensure that the docstrings are at max 80 characters.

**Comment by qiauil** [comment]
Hi, Felix,

The new commit has been pushed; please take a look.

**Comment by Ceyron** [comment]
Thanks for all the work.

I just went over the code for a last time, and I think that the nonconservative eval functions (both for multi-channel and single-channel) should use the `u_hat_dealiased` for creating `nabla_u` in lines 167 and 228 of `_convection.py`. What do you think?

**Comment by qiauil** [comment]
Hi, Felix,

Good point!
To be honest, I am not sure about that. Do you think the linear operation will change the frequency of the field?
Theoretically,  I would prefer to dealias the `nabla_u` as it is the actual one that does the nonlinear operation.

Anyway, I will currently modify the `nabla_u` term with the dealiased value; we could do some experiments later to see the difference.

**Comment by Ceyron** [comment]
The linear operator (i.e., the `derivative_operator`) does not move energy between the modes.

Aliasing occurs when we element-wise multiply two states in state space. To avoid creating any (undesirable) aliases, according to Orszag's 2/3 rule, both factors in the product must be zero-ed out in the upper 1/3 frequency range (which is what `self. dealias` does). Hence, we need to have both `nabla_u` and `u` being dealiased. At least, that's my current understanding of dealiasing. We can talk about it more in a separate thread.

With the modifications, I will now merge the PR. Thanks for the work :). If you are up for it, it would also be cool to extend some of the unit tests for the new four permutations of convection, but I'd say that's a separate PR.

**Pull request merged**